riscv-sbi.adoc

```diff
@@ -0,0 +1,174 @@
+the general RISC-V philosophy of having a small core along with a set of
+optional modular extensions.
+
+### Versioning and History
+
+This document describes a draft of version 0.2 of the RISC-V SBI specification.
+
+#### Changes Since Version 0.1
+
+* The entire v0.1 SBI has been moved to the legacy extension, which is now an
+  optional extension.  This is technically a backwards-incompatible change, 
+
+## Binary Encoding
+
+All SBI functions share a single binary encoding, which facilitates the mixing
+of SBI extensions.  This binary encoding matches the standard RISC-V UNIX
+syscall ABI, which itself is based on the calling convention defined in the
+RISC-V ELF psABI.  In other words, SBI calls are exactly the same as standard
+RISC-V function calls except that:
+
+* An `ecall` is used as the control transfer instruction instead of a `call`
+  instruction.
+* `a7` (or `t0` on RV32E-based systems) encodes the SBI extension ID, which
+  matches how the system call ID is encoded in the standard UNIX system call
+  ABI.
+
+Many SBI extensions also chose to encode an additional function ID in `a6`,
+a scheme similar to the `ioctl()` system call on many UNIX operating systems.
+This allows SBI extensions to encode multiple functions within the space of a
+single extension.
+
+In the name of compatibility, SBI extension IDs and SBI function IDs are
+encoded as signed 32-bit integers.  When passed in registers these follow the
+standard RISC-V calling convention rules.
+
+SBI functions generally return a pair of values in `a0` and `a1`, with `a1`
+returning an error code.  This is analogous to returning the C structure
+
+    struct sbiret {
+        long value;
+        long error;
+    };
+
+Standard SBI error codes are listed below
+
+| Error Type               | Value  |
+| -------------------------|:------:|
+|  SBI_SUCCESS             |  0     |
+|  SBI_ERR_FAILURE         | -1     |
+|  SBI_ERR_NOT_SUPPORTED   | -2     |
+|  SBI_ERR_INVALID_PARAM   | -3     |
+|  SBI_ERR_DENIED          | -4     |
+|  SBI_ERR_INVALID_ADDRESS | -5     |
+
+
+## SBI Base Functionality, Extension ID 0x10
+
+The base of the supervisor binary interface is designed to be as small as
+possible.  As such, it only contains functionality for probing which SBI
+extensions are available and for querying the version of the SBI.  All
+functions in the base must be supported by all SBI implementations, so there
+are no error returns defined.
+
+```
+long sbi_get_sbi_spec_version(void);
+```
+Returns the current SBI specification version.  This function must always
+succeed.  The minor number of the SBI specification is encoded in the low 24
+bits, with the major number encoded in the next 7 bits.  Bit 32 must be 0 and
+is reserved for future expansion.
+
+```
+long sbi_get_sbi_impl_id(void);
+```
+Returns the current SBI implementation ID, which is different for every SBI
+implementation.  It is intended that this implementation ID allows software to
+probe 
```

Comment:
Missing end of sentence.

~~How are implementation IDs assigned to implementations?~~ Perhaps mention that valid implementation IDs are listed below, and how new ones should be requested.